### PR TITLE
Create usaco-1230.mdx

### PR DESCRIPTION
**Pull request by EulerPascal404**
Adding the march silver #1 problem - it is a good application of functional graphs. I will add it to functional graphs module.

_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

**Comment by EulerPascal404** [comment]
What is the problem, would anyone mind helping me with this?

File: solutions/usaco-1230.mdx [hunk 1]
```diff
@@ -0,0 +1,78 @@
+---
+id: usaco-1230
+source: USACO Silver 2022 Open
+title: Redistributing Gifts
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:4]: wrong problem name lol

**Review by envyaims — CHANGES_REQUESTED** [review]
bruh you just copied the official solution -_-

http://www.usaco.org/current/data/sol_prob1_silver_open22.html

review this from https://usaco.guide/general/contributing
image

**Comment by envyaims** [comment]
> Adding the march silver #1 problem - it is a good application of functional graphs. I will add it to functional graphs module.

isnt it already there?

**Comment by EulerPascal404** [comment]
Hey envyaims!

I am a new contributor on the site, so I didn't realize the editorial solutions were not supposed to be added directly to the guide. I have changed the commit with my solutions(and added comments for it to be more helpful). As for the functional graphs module, I meant that the solution was not linked there - but I see that I was wrong, it was just linked directly to the editorial.

Thanks!

File: solutions/usaco-1230.mdx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+This code is similar to the official editorial, but more intuitive and with comments.
+
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:20]:
```suggestion
```
not really needed

File: solutions/usaco-1230.mdx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+This code is similar to the official editorial, but more intuitive and with comments.
+	
+```cpp
+#include <iostream>
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:22]:
```suggestion
```
you're using `bits` already

File: solutions/usaco-1230.mdx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+#include <bits/stdc++.h>
+
+#define int int64_t
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:25]: dear god please no `#define`

File: solutions/usaco-1230.mdx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        z = a[a[z]];
+    }
+    int min_v = v[y];
+    y = a[y];
+    while(y != z){
+        min_v = min(min_v, v[y]);
+        y = a[y];
+    }
+    mark(y);
+    return min_v;
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:54]: tabs

File: solutions/usaco-1230.mdx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+int32_t main()
+{
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:58]: either put the opening brace on the same line or a new line all the time

File: solutions/usaco-1230.mdx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    int n, ans = 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:59]: move `ans` down where you actually use it

File: solutions/usaco-1230.mdx [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    //From each cycle, you can visit all except 1 cow. We make this cow the one
+    //with the lowest v_i. We subtract this from our total answer.
+    for (int i = 0; i < n; i++){
+        if (!visited[i]) ans -= min_in_cycle(i);
+    }
+
+    cout << ans << "\n";
+    return 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:79]:
```suggestion
    cout << ans << endl;
```

File: solutions/usaco-1230.mdx [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+```
+</CPPSection>
+
+<PySection>
+</PySection>
+
+<JavaSection>
+
+</JavaSection>
+
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:91]:
```suggestion
```
what

File: solutions/usaco-1230.mdx [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+#include <iostream>
+#include <bits/stdc++.h>
+
+#define int int64_t
+
+using namespace std;
+
+vector<int> a,v; //cow i wants to visit cow a[i] and gets v[i] points
+vector<vector<int>> graph; //graph[i] stores the cows that want to go to farm i
+vector<bool> visited; //done
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:31]: > done

what this mean

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-1230.mdx (hunk 4 above)
Comment (envyaims) [reply solutions/usaco-1230.mdx:25]: using ll = long long 😎

File: solutions/usaco-1230.mdx [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+//Applies Floyd's algorithm
+int min_in_cycle(int curr){
+    int y = a[curr];
+    int z = a[y];
+    while(y != z){
+        y = a[y];
+        z = a[a[z]];
+    }
+    int min_v = v[y];
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:47]: some comments throughout here would be nice

and better variable names...

File: solutions/usaco-1230.mdx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+vector<vector<int>> graph; //graph[i] stores the cows that want to go to farm i
+vector<bool> visited; //done
+
+void mark(int y){
+    if (visited[y]) return;
+	visited[y] = true;
+	for (int c : graph[y]) mark(c);
+}
+
+//Applies Floyd's algorithm
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:39]: space out these comments like:

```
// text
```

**Review by envyaims — CHANGES_REQUESTED** [review]
overall some more comments would be nice

otherwise it's fine

**Comment by EulerPascal404** [comment]
@envyaims @SansPapyrus683 I think I have made all the requested changes + added more comments. Please merge if it's fine, or if there are other errors, please let me know.

**Comment by envyaims** [comment]
Fix comment spacing for all other comments and use tabs.

**Comment by EulerPascal404** [comment]
I did not know what you meant when adding the tabbing(since it was there in my local IDE). I have fixed it and made the comments uniform.

File: solutions/usaco-1230.mdx [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    	v.resize(n);
+    	a.resize(n);
+    	visited.resize(n);
+    	graph.resize(n);
+
+    	ll ans = 0;
+    	for(int i = 0; i < n; i++){
+        	cin >> a[i] >> v[i];
+        	a[i]--;
+        	graph[a[i]].push_back(i); //since cow i wants to visit farm a[i]
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:70]:
```suggestion
        	graph[a[i]].push_back(i); // cow i wants to visit farm a[i]
```

File: solutions/usaco-1230.mdx [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	if (visited[y]) return;
+	visited[y] = true;
+	for (int c : graph[y]) mark(c);
+}
+
+// applies Floyd's algorithm
+int min_in_cycle(int curr){
+	int y = a[curr];
+	int z = a[y];
+	while(y != z){
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:41]:
```suggestion
	while (y != z) {
```

keep your spacings consistent

File: solutions/usaco-1230.mdx [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+		y = a[y];
+        	z = a[a[z]];
+    	}
+	// y is now an element that is in the cycle
+	int min_v = v[y];
+	y = a[y];
+
+	// traverse the cycle to find the cow with the minimum v_i value
+	while(y != z){
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:50]:
```suggestion
	while (y != z) {
```

File: solutions/usaco-1230.mdx [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+int main(){
+    	int n;
+    	cin >> n;
+    	v.resize(n);
+    	a.resize(n);
+    	visited.resize(n);
+    	graph.resize(n);
+
+    	ll ans = 0;
+    	for(int i = 0; i < n; i++){
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:67]:
```suggestion
    	for (int i = 0; i < n; i++) {
```

File: solutions/usaco-1230.mdx [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        	cin >> a[i] >> v[i];
+        	a[i]--;
+        	graph[a[i]].push_back(i); //since cow i wants to visit farm a[i]
+        	ans += v[i];
+    	}
+
+    	// from each cycle, you can visit all except 1 cow. We make this cow the one
+    	// with the lowest v_i. We subtract this from our total answer.
+    	for (int i = 0; i < n; i++){
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:76]:
```suggestion
    	for (int i = 0; i < n; i++) {
```

**Review by envyaims — CHANGES_REQUESTED** [review]

**Comment by EulerPascal404** [comment]
Committed all the new suggestions.

File: solutions/usaco-1230.mdx [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+typedef long long ll;
+
+vector<int> a,v; // cow i wants to visit cow a[i] and gets v[i] points
```
Comment (envyaims) [inline solutions/usaco-1230.mdx:26]:
```suggestion
vector<int> a, v; // cow i wants to visit cow a[i] and gets v[i] points
```

**Review by envyaims — APPROVED** [review]
great!

**Comment by EulerPascal404** [comment]
I committed the change(although it seems rather like splitting hairs).

**Comment by SansPapyrus683** [comment]
it's best to space out your operators so it's more readable

**Comment by EulerPascal404** [comment]
> it's best to space out your operators so it's more readable

Got it.

File: solutions/usaco-1230.mdx [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+
+vector<int> a, v; // cow i wants to visit cow a[i] and gets v[i] points
+vector<vector<int>> graph; // graph[i] stores the cows that want to go to farm i
+vector<bool> visited; // marks cows as visited once we have processed them
+
+// marks y and the other cows in its other cycle as visited
+void mark(int y){
+	if (visited[y]) return;
+	visited[y] = true;
+	for (int c : graph[y]) mark(c);
```
Comment (SansPapyrus683) [inline solutions/usaco-1230.mdx:34]: braces

**Comment by EulerPascal404** [comment]
Done.

**Review by SansPapyrus683 — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
wait for deployment then i'll merge

**Comment by EulerPascal404** [comment]
Alright, thank you!

**Pull request merged**